NSFS | Concurrency Tests | Warp fails on The specified key does not exist - concurrency = 1000

### Environment info
- NooBaa Version: **noobaa-core-5.17.0-20241012.el9.x86_64**

### Actual behavior

1. Errors from warp versioned with --concurrent=1000:
```
warp: <ERROR> stat error:The specified key does not exist.
warp: <ERROR> stat error:The specified key does not exist.
warp: <ERROR> download error: The specified key does not exist.
warp: <ERROR> stat error:The specified key does not exist.
warp: <ERROR> download error: The specified key does not exist.
warp: <ERROR> download error: The specified key does not exist.
warp: Benchmark data written to "warp-versioned-2024-10-14[124514]-keAl.csv.zst"

Mixed operations.

Operation: DELETE, 10%, Concurrency: 1000, Ran 1h0m1s.
 * Throughput:74.77 obj/s

Operation: GET, 45%, Concurrency: 1000, Ran 1h0m1s.
Errors:10081
 * Throughput:0.32 MiB/s, 333.67 obj/s

Operation: PUT, 15%, Concurrency: 1000, Ran 1h0m1s.
 * Throughput:0.11 MiB/s, 112.10 obj/s

Operation: STAT, 30%, Concurrency: 1000, Ran 1h0m1s.
Errors:6922
 * Throughput:222.40 obj/s

Cluster Total: 0.43 MiB/s, 742.94 obj/s, 17003 errors over 1h0m0s.
Total Errors:17003.
```

### Expected behavior
1. Completed without problems 

### Steps to reproduce

`warp versioned  --host=172.20.100.6{0...9}:6443 --access-key="$access_key" --secret-key="$secret_key"  --obj.size=1k  --duration=1h   --objects=10000 --concurrent=1000  --bucket="bucket50" --insecure --tls`

### More information - Screenshots / Logs / Other output

`Oct 14 13:00:08 c83f2-dan8 [3509675]: [nsfs/3509675] ESC[31m[ERROR]ESC[39m core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resource>/bucket50/Jhl7e1LJ/14.C3I15LHjsQnXi8fk.rnd?versionId=mtime-d4voqndv2cxs-ino-fjecw</Resource><RequestId>m299ff3p-b71c1l-xcx</RequestId></Error> HEAD /bucket50/Jhl7e1LJ/14.C3I15LHjsQnXi8fk.rnd?versionId=mtime-d4voqndv2cxs-ino-fjecw {"host":"172.20.100.61:6443","user-agent":"MinIO (linux; amd64) minio-go/v7.0.66 warp/0.7.7","authorization":"AWS4-HMAC-SHA256 Credential=PylTLnQTwWAPKP71Ipmf/20241014/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=ee6ae569bea2d553a4b5ae8e04757be4dc1efc8df3a890980f33d2e3fae9a0f7","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","x-amz-date":"20241014T171356Z"} Error: No such file or directory - context: Stat _path=/gpfs/remote_rw_fs0/buckets_1/bucket50/Jhl7e1LJ/.versions/14.C3I15LHjsQnXi8fk.rnd_mtime-d4voqndv2cxs-ino-fjecw`

@javieramirez1 looking at the logs you sent on slack I see two main issues:

1. there seems to be files where the version-id in the external attributes doesn't match the files name. this might indicate a corruption:
```
Oct 15 13:10:06 c83f2-dan10 [1639045]: [nsfs/1639045]  [WARN] core.sdk.namespace_fs:: NamespaceFS.read_object_md: retrying retries=1 file_path=/gpfs/remote_rw_fs0/buckets_1/bucket50/zTYkb3SF/.versions/4.e
hrPrMJnZikl5HrE.rnd_mtime-d4wjicrgv3sw-ino-15s2h [Error: No such file or directory] { code: 'ENOENT', context: 'Stat _path=/gpfs/remote_rw_fs0/buckets_1/bucket50/zTYkb3SF/.versions/4.ehrPrMJnZikl5HrE.rnd_
mtime-d4wjicrgv3sw-ino-15s2h ' }
Oct 15 13:10:06 c83f2-dan10 [1639045]: [nsfs/1639045] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exi
st.</Message><Resource>/bucket50/zTYkb3SF/4.ehrPrMJnZikl5HrE.rnd?versionId=mtime-d4wjicrgv3sw-ino-15s2h</Resource><RequestId>m2ap84um-65kshx-thy</RequestId></Error> GET /bucket50/zTYkb3SF/4.ehrPrMJnZikl5H
rE.rnd?versionId=mtime-d4wjicrgv3sw-ino-15s2h {"host":"172.20.100.66:6443","user-agent":"MinIO (linux; amd64) minio-go/v7.0.66 warp/0.7.7","authorization":"AWS4-HMAC-SHA256 Credential=PylTLnQTwWAPKP71Ipmf
/20241015/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=69e6cff9a23d8e9918046b0cf7144990655ffdf856cbcc3d64ded7df71364f47","x-amz-content-sha256":"e3b0c44298fc1c1
49afbf4c8996fb92427ae41e4649b934ca495991b7852b855","x-amz-date":"20241015T172025Z"} Error: No such file or directory - context: Stat _path=/gpfs/remote_rw_fs0/buckets_1/bucket50/zTYkb3SF/.versions/4.ehrPr
```

2. in the end of the log file there is a panic for to many files open:
```
Oct 15 23:55:49 c83f2-dan10 [2771795]: [nsfs/2771795] [ERROR] CONSOLE:: PANIC: process uncaughtException Error: EMFILE: too many open files, uv_resident_set_memory    at process.memoryUsage (node:internal
/process/per_thread:168:5)    at Timeout.memory_monitor [as _onTimeout] (/usr/local/noobaa-core/src/util/panic.js:40:23)    at listOnTimeout (node:internal/timers:573:17)    at process.processTimers (node
```

this might be what causing `Connection closed by foreign host` error

the issue is that the logs file is 12 hours long and its hard to understand which error relates to which issue, for which rpm. also when the errors for the first issue happens it seems that the log level is the lowest, so I am missing a lot of logs.

also if the issue happens after 1 hour, there is no need to run it for 8 hours. can you replicate the issue and send the logs again with full logs, and specify start and end time of the warp. you can set the logs to highest level by running `mms3 config change DEBUGLEVEL="all"` and can you specify the time the watp starts and finishes. also if you use the /var/log/messages logs you can run `tail -f /var/log/messages > logs.txt` before the warp command on the noobaa node, and abort it when the warp finishes. this will store only the logs that were printed during the warp command in the logs.txt file. so the logs file will be much smaller that way

@javieramirez1 
2 questions - 
1. I see you have a space between -- and the stress flag, did you run it without the space? 
> 4)run the next workload addign the flag --stress warp versioned --host=172.20.100.6{0...9}:6443 --access-key="$access_key" --secret-key="$secret_key" --obj.size=1k --duration=1h -- stress --objects=10000 --concurrent=100 --bucket="bucket5001" --insecure –tls
2. do you see in the new logs the PANIC Nadav mentioned in https://github.com/noobaa/noobaa-core/issues/8463#issuecomment-2440100023 ?
```
Oct 15 23:55:49 c83f2-dan10 [2771795]: [nsfs/2771795] [ERROR] CONSOLE:: PANIC: process uncaughtException Error: EMFILE: too many open files, uv_resident_set_memory    at process.memoryUsage (node:internal
/process/per_thread:168:5)    at Timeout.memory_monitor [as _onTimeout] (/usr/local/noobaa-core/src/util/panic.js:40:23)    at listOnTimeout (node:internal/timers:573:17)    at process.processTimers (node
```

on new logs issue is different. we now get ENOENT when getting / heading versions:
```
Oct 28 20:05:09 c83f2-dan10 [3852289]: [nsfs/3852289]  [WARN] core.sdk.namespace_fs:: NamespaceFS.read_object_md: retrying retries=1 file_path=/gpfs/remote_rw_fs0/buckets_2/bucket5001/(T)85WK5/.versions/6
7.lXqtres9RNBuR4Xa.rnd_mtime-d57tem9337k0-ino-t2x5f [Error: No such file or directory] { code: 'ENOENT', context: 'Stat _path=/gpfs/remote_rw_fs0/buckets_2/bucket5001/(T)85WK5/.versions/67.lXqtres9RNBuR4X
a.rnd_mtime-d57tem9337k0-ino-t2x5f ' }
Oct 28 20:05:09 c83f2-dan10 [3852289]: [nsfs/3852289] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resource>/bucket5001/%28T%2985WK5/67.lXqtres9RNBuR4Xa.rnd?versionId=mtime-d57tem9337k0-ino-t2x5f</Resource><RequestId>m2torxie-bpipd9-chr</RequestId></Error> HEAD /bucket5001/%28T%2985WK5/67.lXqtres9RNBuR4Xa.rnd?versionId=mtime-d57tem9337k0-ino-t2x5f {"host":"172.20.100.67:6443","user-agent":"MinIO (linux; amd64) minio-go/v7.0.66 warp/0.7.7","authorization":"AWS4-HMAC-SHA256 Credential=38MmE379Nxz54ELdjjUa/20241029/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=80af6395eae063609c5f37f5b141e9ed0843e4565ba856156154e6df29e9cf2b","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","x-amz-date":"20241029T001449Z"} Error: No such file or directory - context: Stat _path=/gpfs/remote_rw_fs0/buckets_2/bucket5001/(T)85WK5/.versions/67.lXqtres9RNBuR4Xa.rnd_mtime-d57tem9337k0-ino-t2x5f
```
this indicates that the version is missing completely. issue was reproduced on my enviorment.
the issue happens when the version we try to delete is the latest version. and the latest version is replaced before we delete it. then we delete the wrong version. added a check that the version we try to delete is the intended one.

ran warp on my system for 30 min with 1000 concurrency,  there were no issues:

```
[root@tmtscalets-master warp]# ./warp versioned --host=10.11.87.62:6443 --access-key="3MwC7LPBgfQ1maEg4ZLN" --secret-key="auCOhJrI6znaabNZv3oNr4d4QltxWg3Z/sJodY1H" --obj.size=1k --concurrent=1000 --duration=30m --bucket=warp-bu5ket1 --insecure --tls --objects=10000
warp: Benchmark data written to "warp-versioned-2024-10-29[050626]-34vw.csv.zst"                                                                                                                            
Mixed operations.
Operation: DELETE, 10%, Concurrency: 1000, Ran 28m38s.
 * Throughput: 15.33 obj/s

Operation: GET, 45%, Concurrency: 1000, Ran 28m38s.
 * Throughput: 0.07 MiB/s, 68.98 obj/s

Operation: PUT, 15%, Concurrency: 1000, Ran 28m38s.
 * Throughput: 0.02 MiB/s, 22.89 obj/s

Operation: STAT, 30%, Concurrency: 1000, Ran 28m38s.
 * Throughput: 46.03 obj/s

Cluster Total: 0.09 MiB/s, 153.09 obj/s over 28m38s.
warp: Cleanup Done.[root@tmtscalets-master warp]#                                                                                                                                                           
```

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -963,6 +963,7 @@ class NamespaceFS {
                     dbg.warn(`NamespaceFS.read_object_md: retrying retries=${retries} file_path=${file_path}`, err);
                     retries -= 1;
                     if (retries <= 0 || !native_fs_utils.should_retry_link_unlink(is_gpfs, err)) throw err;
+                    await P.delay(get_random_delay(config.NSFS_RANDOM_DELAY_BASE, 0, 50));
                 }
             }
             this._throw_if_delete_marker(stat, params);
@@ -1037,6 +1038,7 @@ class NamespaceFS {
                             {bucket_path: this.bucket_path, object_name: params.key}, err);
                         throw err;
                     }
+                    await P.delay(get_random_delay(config.NSFS_RANDOM_DELAY_BASE, 0, 50));
                 }
             }
             this._throw_if_delete_marker(stat, params);
@@ -2847,6 +2849,13 @@ class NamespaceFS {
                 const deleted_latest = file_path === latest_version_path;
                 if (deleted_latest) {
                     gpfs_options = await this._open_files_gpfs(fs_context, file_path, undefined, undefined, undefined, undefined, true);
+                    if (gpfs_options) {
+                        const src_stat = await gpfs_options.delete_version.src_file.stat(fs_context);
+                        if (this._is_mismatch_version_id(src_stat, version_id)) {
+                            dbg.warn('NamespaceFS._delete_single_object_versioned mismatch version_id', file_path, version_id, this._get_version_id_by_xattr(src_stat));
+                            throw error_utils.new_error_code('MISMATCH_VERSION', 'file version does not match the version we asked for');
+                        }
+                    }
                     const bucket_tmp_dir_path = this.get_bucket_tmpdir_full_path();
                     await native_fs_utils.safe_unlink(fs_context, file_path, version_info,
                         gpfs_options?.delete_version, bucket_tmp_dir_path);
@@ -3007,15 +3016,17 @@ class NamespaceFS {
         let latest_ver_info;
         for (;;) {
             try {
-                // get latest version_id if exists
+                // TODO get latest version from file in POSIX like in GPFS path
                 latest_ver_info = await this._get_version_info(fs_context, latest_ver_path);
-                const versioned_path = latest_ver_info && this._get_version_path(params.key, latest_ver_info.version_id_str);
-                const versioned_info = latest_ver_info && await this._get_version_info(fs_context, versioned_path);
-
-                dbg.log1('Namespace_fs._delete_latest_version:', latest_ver_info, versioned_path, versioned_info);
+                dbg.log1('Namespace_fs._delete_latest_version:', latest_ver_info);
                 if (latest_ver_info) {
+                    if (is_gpfs) {
                     gpfs_options = await this._open_files_gpfs(fs_context, latest_ver_path, undefined, undefined, undefined,
-                        undefined, true, versioned_info);
+                            undefined, true);
+                        const latest_fd = gpfs_options?.move_to_dst?.dst_file;
+                        latest_ver_info = latest_fd && await this._get_version_info(fs_context, undefined, latest_fd);
+                    }
+                    const versioned_path = latest_ver_info && this._get_version_path(params.key, latest_ver_info.version_id_str);
 
                     const suspended_and_latest_is_not_null = this._is_versioning_suspended() &&
                         latest_ver_info.version_id_str !== NULL_VERSION_ID;
@@ -3172,7 +3183,7 @@ class NamespaceFS {
         let dir_file;
         let versioned_file;
         try {
-            // open /versions/key_ver file if exists
+            // open /versions/key_ver file if exists. TODO is versioned_file needed
             versioned_file = versioned_info && await native_fs_utils.open_file(fs_context, this.bucket_path, versioned_info.path, 'r');
 
             // open files for deletion flow

